kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/AwsV4SigningHttpRequestBuilder.java

```diff
@@ -0,0 +1,364 @@
+        return this;
+    }
+
+    @Override
+    public Builder method(String method, BodyPublisher bodyPublisher) {
+        builder.method(method, digestingPublisher(bodyPublisher));
+        return this;
+    }
+
+    @Override
+    public Builder copy() {
+        return new AwsV4SigningHttpRequestBuilder(accessKey, secretKey, region, service, date, builder.copy());
+    }
+
+    @Override
+    public Builder uri(URI uri) {
+        builder.uri(uri);
+        return this;
+    }
+
+    @Override
+    public HttpRequest build() {
+        signRequest();
+        return builder.build();
+    }
+
+    @NonNull
+    private BodyPublisher digestingPublisher(BodyPublisher bodyPublisher) {
+        var items = new ArrayList<BodyPublisher>();
+
+        bodyPublisher.subscribe(new Flow.Subscriber<>() {
+            final MessageDigest digest = getSha256Digester();
+
+            @Override
+            public void onSubscribe(Flow.Subscription subscription) {
+                subscription.request(Long.MAX_VALUE);
+            }
+
+            @Override
+            public void onNext(ByteBuffer item) {
+                digest.update(item.array());
+                items.add(HttpRequest.BodyPublishers.ofByteArray(item.array()));
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                throw new IllegalStateException(throwable);
+            }
+
+            @Override
+            public void onComplete() {
+                payloadHexedSha56 = HEX_FORMATTER.formatHex(digest.digest());
+            }
+        });
+
+        return HttpRequest.BodyPublishers.concat(items.toArray(new BodyPublisher[]{}));
+    }
+
+    /**
+     * This method implements <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html">signs the request</a>
+     * using the AWS v4 algorithm.
+     */
+    private void signRequest() {
+        var isoDateTime = DATE_TIME_FORMATTER.format(date);
+        var isoDate = isoDateTime.substring(0, 8);
+        var unsignedRequest = builder.build();
+
+        var allHeaders = new HashMap<String, String>();
+        unsignedRequest.headers().map().keySet().forEach(k -> unsignedRequest.headers().firstValue(k).ifPresent(value -> allHeaders.put(k, value)));
+        allHeaders.put(HOST_HEADER, getHostHeaderForSigning(unsignedRequest.uri()));
+        allHeaders.put(X_AMZ_DATE_HEADER, isoDateTime);
+
+        try {
+            var canonicalRequest = computeCanonicalRequest(allHeaders, unsignedRequest);
+            var stringToSign = computeStringToSign(canonicalRequest, isoDateTime, isoDate);
+            var authorization = computeAuthorization(canonicalRequest, stringToSign, isoDate);
+
+            builder.header(AUTHORIZATION_HEADER, authorization);
+            builder.header(X_AMZ_DATE_HEADER, isoDateTime);
+        }
+        catch (Exception e) {
+            throw new IllegalStateException("Failed to sign the request", e);
```

Comment:
This doesn't look like an ISE to me, because you don't know what invariant is being violated in this case. Better to either:
* reuse `KmsException` from our Kms API, or
* declare your own `AwsKmsException extends RuntimeException`